File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.js
function solution(enroll, referral, seller, amount) {
  const enrollInfo = {};

  for (let i = 0; i < enroll.length; i += 1) {
    enrollInfo[enroll[i]] = { parent: referral[i], value: 0 };
  }

  for (let i = 0; i < seller.length; i += 1) {
    const sel = seller[i];
    const am = amount[i];

    const sellMoney = 100 * am;
    let parentMoney = Math.floor(sellMoney * 0.1);
    let myMoney = sellMoney - parentMoney;
    if (myMoney === 0) {
      myMoney = parentMoney;
      parentMoney = 0;
    }
    enrollInfo[sel].value += myMoney;
    let parent = enrollInfo[sel].parent;
    while (true) {
      if (parent === "-" || parentMoney === 0) break;
      const childMoney = parentMoney;
      parentMoney = Math.floor(childMoney * 0.1);
      myMoney = childMoney - parentMoney;

      if (myMoney === 0) {
        myMoney = parentMoney;
        parentMoney = 0;
      }
      enrollInfo[parent].value += myMoney;
      parent = enrollInfo[parent].parent;
    }
  }
  const result = Object.entries(enrollInfo).map(v=>v[1].value);
  return result
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;

class Solution {
    public static class Seller {
        String name;
        int money;

        public Seller(String name, int money) {
            this.name = name;
            this.money = money;
        }

        @Override
        public String toString() {
            return "Seller{" +
                    "name='" + name + '\'' +
                    ", money=" + money +
                    '}';
        }
    }

    private Map<String, Seller> sellers = new HashMap<>();

    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount){
        init(enroll, referral);

        for (int i = 0; i < seller.length; i++) {
            String me = seller[i];
            int money = amount[i] * 100;

            while(sellers.containsKey(me)) {
                String parent = sellers.get(me).name;
                int parentMoney = calc(money);
                int myMoney = money - parentMoney;
                sellers.get(me).money += myMoney;

                me = parent;
                money = parentMoney;
                if(money == 0) break;
            }
        }

        int[] answer = new int[enroll.length];
        for (int i = 0; i < enroll.length; i++) {
            answer[i] = sellers.get(enroll[i]).money;
        }

        return answer;
    }

    public void init(String[] enroll, String[] referral) {
        for (int i = 0; i < enroll.length; i++) {
            sellers.put(enroll[i], new Seller(referral[i], 0));
        }
    }

    public int calc(int price) {
        // 10%를 계산
        return (int)(price*0.1);
    }
}